### ISSUE TITLE
Data binding override dirty check not working when using linkPaths

### ISSUE THREAD

**Issue by elonmallin** [Polymer/polymer#3945]
The recommended way to [override dirty check](https://www.polymer-project.org/1.0/docs/devguide/model-data#override-dirty-check) on a data binding is not working:
`var tmp = this.myArray;
this.myArray = [];
this.myArray = tmp;`

The problem occurs in conjunction with `linkPaths`.
Workarounds include `setTimeout` and strange `notifyPath` combination.
## Repro

Click the 'Edit' button.
Change the name.
Click back and see the name is not updated.
Check the `_refresh` function and uncomment one of the two workarounds to see it work.
### Live Demo

https://jsbin.com/jigipahaco/2/edit?html,console,output
## Versions

Polymer 1.6.1

**Comment by arthurevans** [comment]
It appears that it also works if you _reverse the order of the arguments to `linkPaths`_:

https://jsbin.com/kenijel/1/edit?html,console,output

```
// Works
this.people = this.data[0].items;
this.linkPaths('data.#0.items', 'people');
// Does not work
this.people = this.data[0].items;
this.linkPaths( 'people', 'data.#0.items');
```

There's nothing in the API docs suggesting that the argument order is significant or (if it is) which argument should go first. I thought since the path mapping should work both ways, order wouldn't be significant here, but apparently it is.

**Comment by elonmallin** [comment]
The reverse order makes no difference as far as I can tell. The important thing to notice is that I `unlinkPaths` when we switch to `isEdit=true` and link it back up again when we switch back. This is why a override dirty check is needed in the first place.

#### jsbin WITH unlinkPath with reverse order arguments
https://jsbin.com/voleweyumu/1/edit?html,console,output

**Comment by arthurevans** [comment]
Ah, I see. I didn't switch the order in the unlinkPaths call, so the path wasn't getting unlinked. (So in answer to my own question, it appears that the order matters in as much as `unlinkPaths(a, b)` won't unlink paths if they were originally linked as `linkPaths(b, a)`.) 

Out of curiosity, why are you unlinking the paths? If you remove the `unlinkPaths` call, it resolves the problem. It would make sense to call `unlinkPaths` if you were replacing the entire `this.person` object, but that's not the case here.

Alternately, if you move the `this.people = []` call to just before the `unlinkPaths` call, it also works fine.

https://jsbin.com/yiwiqu/2/edit?html,console,output

**Comment by elonmallin** [comment]
Yep the order seems to be relevant in that way.

The only reason I am doing `unlinkPaths` is because it "feels" right. If we imagine a potentially bigger app with more complex data, I'm switching to a different view so I no longer need updates to reflect in the UI (but maybe this is just unnecessary pre-optimization). And it's not guaranteed that the next time I show `this.people` it is the same people as before. It might be the people of another department etc if that makes sense. So the flow of the example is not really same as my app.

The reason I didn't also just do `this.people = []` with the `unlinkPaths` like in your example now is because I'm switching views with an animation so I still need the data there until the animation is done. So I figured it's just easier to do it when I do `linkPaths` again.

#### TLDR
But to narrow down the problem. The reason it works in your example now is not because you call `this.people = []` before the `unlinkPaths`, it works because it doesn't happen in the same frame is what I think. I suspect the issue has something to do with "debounce time (in polymer speak?)". Hence the `setTimeout` workaround in the original jsbin.

**Comment by jfrazzano** [comment]
can i get a re-link to the bin, this sounds cool.  Thanks

**Comment by elonmallin** [comment]
https://jsbin.com/kazesodenu/1/edit?html,console,output
This is the bin with the `setTimout` used so it works. Edit the `_refresh` function comments to see the problem.
Or use the original bin: https://jsbin.com/jigipahaco/2/edit?html,console,output

**Comment by jfrazzano** [comment]
I did this…   _refresh: function() {
            
            this.set("people",this.data[0].items);

and it seemed to work in the original bin, I also tried a version of  the below, which worked one run, on a nonsense value, then I was gonna see if I could link it to the input and data property through the setter, but I hurled a period somewhere on the page, and then was unable to locate and remove the period i accidentally clicked—been working on some code way to long.  Anyway I hope it helps.  Top definitely works.  Bottom can be implemented in constructors or subsequently, even off thread, to set up your own zero thread cost binding/observer/listener… its kinda cool.  If you have a use case I could show adapt it if you are interested and its not too involved.  Have a good night and thanks for throwing the problem back to me. It was cool.

sincerely jason

Object.defineProperty(this, "dataArrays", {
        enumerable: true,
        get : function ( ) {
            return bindTo.value;
        },
        set : function ( val ) {
            bindTo.value = val;
        }
            });

**Comment by elonmallin** [comment]
That doesn't fix the problem at least in my browser. `this.set('people', this.data[0].items)` is still the same problem as `this.people = this.data[0].items`.
https://jsbin.com/zegimewega/1/edit?html,console,output

The other part sounds more like a tangent not related to the bug but thanks for looking in to it.
